src/thinblock.cpp

```diff
@@ -380,58 +335,109 @@ bool CXThinBlockTx::HandleMessage(CDataStream &vRecv, CNode *pfrom)
-    std::map<uint64_t, CTransaction> mapMissingTx;
-    BOOST_FOREACH (CTransaction tx, thinBlockTx.vMissingTx)
-        mapMissingTx[tx.GetHash().GetCheapHash()] = tx;
+    BOOST_FOREACH (const CTransaction tx, thinBlockTx.vMissingTx)
+        pfrom->mapMissingTx[tx.GetHash().GetCheapHash()] = tx;
 
+    // Get the full hashes from the xblocktx and add them to the thinBlockHashes vector.  These should
+    // be all the missing or null hashes that we re-requested.
     int count = 0;
-    uint64_t maxAllowedSize = maxMessageSizeMultiplier * excessiveBlockSize;
-    CTransaction nulltx;
-    uint64_t nSizeNullTx = RecursiveDynamicUsage(nulltx);
-    for (size_t i = 0; i < pfrom->thinBlock.vtx.size(); i++)
+    for (size_t i = 0; i < pfrom->thinBlockHashes.size(); i++)
     {
-        if (pfrom->thinBlock.vtx[i].IsNull())
+        if (pfrom->thinBlockHashes[i].IsNull())
         {
-            std::map<uint64_t, CTransaction>::iterator val = mapMissingTx.find(pfrom->xThinBlockHashes[i]);
-            if (val != mapMissingTx.end())
+            std::map<uint64_t, CTransaction>::iterator val = pfrom->mapMissingTx.find(pfrom->xThinBlockHashes[i]);
+            if (val != pfrom->mapMissingTx.end())
             {
-                pfrom->thinBlock.vtx[i] = val->second;
-                pfrom->thinBlockWaitingForTxns--;
-
-                // In order to prevent a memory exhaustion attack we track transaction bytes used to create Block
-                // to see if we've exceeded any limits and if so clear out data and return.
-                uint64_t nTxSize = RecursiveDynamicUsage(val->second) - nSizeNullTx;
-                if (thindata.AddThinBlockBytes(nTxSize, pfrom) > maxAllowedSize)
-                {
-                    if (ClearLargestThinBlockAndDisconnect(pfrom))
-                        return error("xthin block has exceeded memory limits of %ld bytes", maxAllowedSize);
-                }
+                pfrom->thinBlockHashes[i] = val->second.GetHash();
             }
             count++;
         }
     }
     LogPrint("thin", "Got %d Re-requested txs, needed %d of them\n", thinBlockTx.vMissingTx.size(), count);
 
-    if (pfrom->thinBlockWaitingForTxns == 0)
+    // At this point we should have all the full hashes in the block. Check that the merkle
+    // root in the block header matches the merkel root calculated from the hashes provided.
+    bool mutated;
+    uint256 merkleroot = ComputeMerkleRoot(pfrom->thinBlockHashes, &mutated);
+    if (pfrom->thinBlock.hashMerkleRoot != merkleroot || mutated)
+    {
+        thindata.ClearThinBlockData(pfrom);
+        {
+            LOCK(pfrom->cs_mapthinblocksinflight);
+            pfrom->mapThinBlocksInFlight.erase(inv.hash);
+        }
+        LOCK(cs_main);
+        Misbehaving(pfrom->GetId(), 100);
+        return error("Thinblock merkle root does not match computed merkle root, peer=%d", pfrom->GetId());
+    }
+    LogPrint("thin", "Merkle Root check passed\n");
```

Comment:
Similarly please add xthin in the log message.  The logs are very confusing at the moment as they are sometimes unattributed or say thin when they mean xthin.  As activity is bursty and overlapping it's hard to understand what's happening.  Perhaps mention the peer too.